src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -225,28 +228,39 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
 
         public IList<SiloAddress> GetCompatibleSilos(PlacementTarget target)
         {
-            // For test only: if we have silos that are not yet in the Cluster TypeMap, we assume that they are compatible
-            // with the current silo
-            if (this.config.AssumeHomogenousSilosForTesting)
-                return AllActiveSilos;
-
-            var typeCode = target.GrainIdentity.TypeCode;
-            IReadOnlyList<SiloAddress> silos;
-            if (target.InterfaceVersion > 0)
-            {
-                var version = GrainTypeManager.GetAvailableVersions(target.InterfaceId).Max();
-                silos = GrainTypeManager.GetSupportedSilos(typeCode, target.InterfaceId, version);
-            }
-            else
+            var key = Tuple.Create(target.GrainIdentity.TypeCode, target.InterfaceId, target.InterfaceVersion);
+
+            return this.compatibleSilosCache.GetOrAdd(key, tuple =>
```

Comment:
This needlessly allocates a closure to capture `this`. It would be preferable to either use `TryGet` or to allocate the closure once in the constructor and keep it in a field.